# Яндекс.Музыка

Андрей, привет! 👋

Меня зовут Артем Пиляй. Я буду проверять твой проект. Со мной можно на «ты».
 
В тетрадке ты найдешь комментарии к проекту, они будут в отдельных ячейках Markdown с заголовком «Ревьюер», если проверок будет несколько, то «Ревьюер v_».  
Твои проекты станут еще круче, если ты воспользуешься комментариями в дальнейших проектах.
 
На заметку: Повторная проверка пройдет быстрее если ты не будешь перемещать, изменять и удалять мои комментарии. Еще ускорить повторную проверку можно если ты будешь писать о внесенных изменениях в синих ячейках Markdown, заготовку я тебе приготовил ниже. =)


<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Такой комментарий нужно исправить обязательно, он критически влияет на удачное выполнение проекта.

</div>

<div class="alert alert-warning">

<b>Ревьюер: 👉</b> Такой комментарий является рекомендацией или советом. Можешь использовать их на своё усмотрение.

</div>

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Такой комментарий в начале проекта означает, что работа принята. В основной части работы комментарии такого цвета говорят о том, что было сделано что-то качественное и правильное. 😉

</div>

<div class="alert alert-info">

<b>Студент:</b> Такой комментарий можно будет быстро найти.

</div>

---

<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Андрей, из-за того, что в датафрейме (таблице) остались дубликаты, в дальнейших вычислениях будут ошибки. Исправь их, пожалуйста. 
    
Будет здорово, если задашь интересующие тебя вопросы. Буду рад помочь! 🙂

</div>

<div class="alert alert-danger">

<b>Ревьюер v2:</b> Проект еще не принят, осталось исправить несколько ошибок.
    
Жду правок 🙂

</div>

<div class="alert alert-danger">

<b>Ревьюер v3:</b> Андрей, осталось чуть-чуть и проект будет принят!
    
Жду правок 😉

</div>

<div class="alert alert-success">

<b>Ревьюер v4: ✌️</b> Андрей, упорства тебе не занимать, ты крут! А это залог успеха. 
    
Только увидел, что ты оставлял комментарии в коде. На вопросы ответил, и твой ответ на вопрос нашел. На будущее оставляй комментарии так чтобы их было видно.

Желаю тебе успехов в дальнейшем обучении и во всех твоих начинаниях, поздравляю **проект принят**! 🥳 🎉

</div>

Сравнение Москвы и Петербурга окружено мифами. Например:
 * Москва — мегаполис, подчинённый жёсткому ритму рабочей недели;
 * Петербург — культурная столица, со своими вкусами.

На данных Яндекс.Музыки вы сравните поведение пользователей двух столиц.

**Цель исследования** — проверьте три гипотезы:
1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.
2. В понедельник утром в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города. 
3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

**Ход исследования**

Данные о поведении пользователей вы получите из файла `yandex_music_project.csv`. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных. 

Вы проверите данные на ошибки и оцените их влияние на исследование. Затем, на этапе предобработки вы поищете возможность исправить самые критичные ошибки данных.
 
Таким образом, исследование пройдёт в три этапа:
 1. Обзор данных.
 2. Предобработка данных.
 3. Проверка гипотез.



## Обзор данных

Составьте первое представление о данных Яндекс.Музыки.




Основной инструмент аналитика — `pandas`. Импортируйте эту библиотеку.

In [1]:
# импорт библиотеки pandas
import pandas as pd

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Молодец, что используешь сокращение `pd` для библиотеки Pandas. Это общепринятое сокращение для этой библиотеки, и делает твой код более понятным для другого аналитика. 

</div>

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Верное решенье оставлять комментарий над кодом. Во-первых, его так удобнее читать, во-вторых, он не останется за пределами ячейки. 

</div>

Прочитайте файл `yandex_music_project.csv` из папки `/datasets` и сохраните его в переменной `df`:

In [2]:
# чтение файла с данными и сохранение в df
df = pd.read_csv('/datasets/yandex_music_project.csv')

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Этот шаг тоже выполнен правильно. Важно указать верный путь к файлу с данными.
* На заметку: 
    * Слэш в начале пути указывает, что нужно искать файл в папке.
    * Если будешь выполнять проекты локально, не забывай проверять их работоспособность после заливки в JupyterHub, это сэкономит твое время.

</div>

Выведите на экран первые десять строк таблицы:

In [3]:
# получение первых 10 строк таблицы df
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

             City        time        Day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Теперь мы познакомились с данными и сделали это аккуратно, осмотрев первые десять строк. Метод `print()` или `display()` ненужно использовать, когда код находится в последней строчке ячейки, Юпитер выведет информацию и без них.
* На заметку: Посмотреть «хвост» можно методом `tail()`, а вывести в случайном порядке можно методом `sample()`.

</div>

Одной командой получить общую информацию о таблице:

In [4]:
# получение общей информации о данных в таблице df
df.info()



#столбец userID необходимо написать в змеином регистре
#  я не понимаю как и где нужно описать


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
  userID    65079 non-null object
Track       63848 non-null object
artist      57876 non-null object
genre       63881 non-null object
  City      65079 non-null object
time        65079 non-null object
Day         65079 non-null object
dtypes: object(7)
memory usage: 3.5+ MB


<div class="alert alert-success">

<b>Ревьюер: 👍</b> Хорошо. Метод `info()`, как и следует из названия очень информативен. 🙂 ~`display()`~

</div>

Итак, в таблице семь столбцов. Тип данных во всех столбцах — `object`.

Согласно документации к данным:
* `userID` — идентификатор пользователя;
* `Track` — название трека;  
* `artist` — имя исполнителя;
* `genre` — название жанра;
* `City` — город пользователя;
* `time` — время начала прослушивания;
* `Day` — день недели.

В названиях колонок видны три нарушения стиля:
1. Строчные буквы сочетаются с прописными.
2. Встречаются пробелы.
3. Найдите ещё одну проблему в названии колонок и опишите её в этом пункте.



Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


<a id="error_1"></a> 
<div class="alert alert-danger">

<b>Ревьюер: ✋</b> В предыдущей ячейке надо ответить на третий вопрос.

</div>

<a id="error_2_1"></a> 
<div class="alert alert-danger">

<b>Ревьюер v2: ✋</b> В предыдущей ячейке надо ответить на третий вопрос.

</div>

<a id="error_3_1"></a> 
<div class="alert alert-danger">

<b>Ревьюер v3: ✋</b> Ответь на вопрос: "Найдите ещё одну проблему в названии колонок и опишите её в этом пункте." 
    
Подсказка: 🐪 или 🐍?

</div>

<div class="alert alert-success">

<b>Ревьюер v4: 👍</b> Оставляй пожалуйста комментарии так, чтобы их было видно. Пробежался глазами по всем комментариям в коде и нашел ответ на вопрос. Просто видел, что ты стараешься и все делаешь хорошо и мне было непонятно почему ты игнорируешь этот пункт. 🙂

</div>

**Выводы**

В каждой строке таблицы — данные о прослушанном треке. Часть колонок описывает саму композицию: название, исполнителя и жанр. Остальные данные рассказывают о пользователе: из какого он города, когда он слушал музыку. 

Предварительно можно утверждать, что, данных достаточно для проверки гипотез. Но встречаются пропуски в данных, а в названиях колонок — расхождения с хорошим стилем.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных
Исправьте стиль в заголовках столбцов, исключите пропуски. Затем проверьте данные на дубликаты.

### Стиль заголовков
Выведите на экран названия столбцов:

In [5]:
# перечень названий столбцов таблицы df
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

<div class="alert alert-warning">

<b>Ревьюер: 👉</b> Метод `columns` выбран верно, но функция `display()` здесь не нужна, потому что это последняя/единственная строка. Поэтому Юпитер выведет информацию без использования функций `display()` или `print()`.

</div>

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Классно, что ты исправляешь не только «красные» комментарии, но и «желтые»!

</div>

Приведите названия в соответствие с хорошим стилем:
* несколько слов в названии запишите в «змеином_регистре»,
* все символы сделайте строчными,
* устраните пробелы.

Для этого переименуйте колонки так:
* `'  userID'` → `'user_id'`;
* `'Track'` → `'track'`;
* `'  City  '` → `'city'`;
* `'Day'` → `'day'`.

In [6]:
# переименование столбцов
df = df.rename(columns={'  userID': 'user_id', 'Track': 'track', '  City  ': 'city', 'Day': 'day'})

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Молодец! Верно, выбран метод переименования. Он удобен тем, что сработает без ошибок, даже если изменится порядок столбцов.

</div>

Проверьте результат. Для этого ещё раз выведите на экран названия столбцов:

In [7]:
# проверка результатов - перечень названий столбцов
df.columns
# когда вижу слово выведите автоматически пишу display

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

<div class="alert alert-warning">

<b>Ревьюер: 👉</b> Здесь тоже не нужна функция `display()`, только  `df.columns` и все. 🙂

</div>

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Вижу, молодец!

</div>

### Пропуски значений
Сначала посчитайте, сколько в таблице пропущенных значений. Для этого достаточно двух методов `pandas`:

In [8]:
# подсчёт пропусков
df.isna().sum()

user_id       0
track      1231
artist     7203
genre      1198
city          0
time          0
day           0
dtype: int64

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Все правильно.

</div>

Не все пропущенные значения влияют на исследование. Так в `track` и `artist` пропуски не важны для вашей работы. Достаточно заменить их явными обозначениями.

Но пропуски в `genre` могут помешать сравнению музыкальных вкусов в Москве и Санкт-Петербурге. На практике было бы правильно установить причину пропусков и восстановить данные. Такой возможности нет в учебном проекте. Придётся:
* заполнить и эти пропуски явными обозначениями,
* оценить, насколько они повредят расчётам. 

Замените пропущенные значения в столбцах `track`, `artist` и `genre` на строку `'unknown'`. Для этого создайте список `columns_to_replace`, переберите его элементы циклом `for` и для каждого столбца выполните замену пропущенных значений:

In [9]:
# перебор названий столбцов в цикле и замена пропущенных значений на 'unknown'

columns_to_replace = ['track', 'artist','genre']

for i in columns_to_replace:
    df[i] = df[i].fillna('unknown') 


<div class="alert alert-success">

<b>Ревьюер: 👍</b> Круто, что ты формируешь список, а затем используешь его в цикле `for`. Такой код удобно читать, за счет списка и он не занимает много строк, за счет цикла.

</div>

Убедитесь, что в таблице не осталось пропусков. Для этого ещё раз посчитайте пропущенные значения.

In [10]:
# подсчёт пропусков
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Верно. На самом деле, `isnull()` и `isna()` делают одно и то же, но использование `isna()` предпочтительнее, потому что `isnull()` – это псевдоним `isna()`.

</div>

### Дубликаты
Посчитайте явные дубликаты в таблице одной командой:

In [11]:
# подсчёт явных дубликатов
df.duplicated().sum()

3826

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Хорошо. Ты используешь оптимальный метод для подсчета дубликатов.

</div>

Вызовите специальный метод `pandas`, чтобы удалить явные дубликаты:

In [12]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df = df.drop_duplicates().reset_index(drop = True)
# я видел, что метод не сработал, но хотел потом вернуться к этому моменту и забыл. Спасибо!

<a id="error_2"></a> 
<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Методы использованы правильно, но нужно было сохранить обновленный датафрейм. Посмотри, в следующей ячейки ты делаешь проверку на дубликаты, результатом должен быть `0`. Подсказка:

```py
data.loc[data[‘column’] == value] # вывод на экран
data  = data.loc[data[‘column’] == value] # присваивание нового значения датафрейму
```

</div>

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Все верно.

</div>

Ещё раз посчитайте явные дубликаты в таблице — убедитесь, что полностью от них избавились:

In [13]:
# проверка на отсутствие дубликатов
df.duplicated().sum()

0

Теперь избавьтесь от неявных дубликатов в колонке `genre`. Например, название одного и того же жанра может быть записано немного по-разному. Такие ошибки тоже повлияют на результат исследования.

Выведите на экран список уникальных названий жанров, отсортированный в алфавитном порядке. Для этого:
* извлеките нужный столбец датафрейма, 
* примените к нему метод сортировки,
* для отсортированного столбца вызовите метод, который вернёт уникальные значения из столбца.

In [14]:
# Просмотр уникальных названий жанров
df['genre'].sort_values().unique()
# не заметил что сперва нужно было отсортировать. Исправлено.

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

<a id="error_3"></a> 
<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Здесь нужно было применить метод, сортирующий значения. Подсказка:
    
Для сортировки датафрейма используется метод ` sort_values()`.

</div>

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Супер.

</div>

Просмотрите список и найдите неявные дубликаты названия `hiphop`. Это могут быть названия с ошибками или альтернативные названия того же жанра.

Вы увидите следующие неявные дубликаты:
* *hip*,
* *hop*,
* *hip-hop*.

Чтобы очистить от них таблицу, напишите функцию `replace_wrong_genres()` с двумя параметрами: 
* `wrong_genres` — список дубликатов,
* `correct_genre` — строка с правильным значением.

Функция должна исправить колонку `genre` в таблице `df`: заменить каждое значение из списка `wrong_genres` на значение из `correct_genre`.

In [15]:
# Функция для замены неявных дубликатов

def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre']=df['genre'].replace(wrong_genres, correct_genre) 
# внимательность не мой конек (. Подскажи почему цикл работал не смотря на неправильную переменную?
# Исправлено!


<div class="alert alert-success">

<b>Ревьюер v4: 👍</b> О, нашел вопрос!

Цикл работал некорректно, но работал потому что `wrong_genres` конечно были в `wrong_genres`, ведь это одно и тоже. Например, `а` есть в `абвгд` и `абвгд` есть в `абвгд`.


</div>

<a id="error_4"></a> 
<div class="alert alert-danger">

<b>Ревьюер: ✋</b> В цикле есть ошибка, смотри:

```py
...
for значение in значениях:
...
```

</div>

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Ты понял в чем ошибка, отлично.

</div>

<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Андрей, из-за того, что в датафрейме (таблице) остались дубликаты, в дальнейших вычислениях будут ошибки. Исправь критические ошибки и продолжим исследование. Для удобства поиска я добавил тебе ссылки.

* [Ответь на вопрос](#error_1)
* [Сохрани обновленный датафрейм](#error_2)
* [Отсортируй уникальные значения](#error_3)
* [Переименуй переменную в цикле](#error_4)
    
Будет здорово, если задашь интересующие тебя вопросы. Буду рад помочь!

</div>


Вызовите `replace_wrong_genres()` и передайте ей такие аргументы, чтобы она устранила неявные дубликаты: вместо `hip`, `hop` и `hip-hop` в таблице должно быть значение `hiphop`:

In [16]:
# Устранение неявных дубликатов
wrong_genres = ['hip', 'hop','hip-hop']
correct_genre = 'hiphop'
replace_wrong_genres(wrong_genres, correct_genre)

Проверьте, что заменили неправильные названия:

*   hip
*   hop
*   hip-hop

Выведите отсортированный список уникальных значений столбца `genre`:

In [17]:
# Проверка на неявные дубликаты
df['genre'].sort_values().unique()

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Нет дубликатов – нет проблем.

</div>

**Выводы**

Предобработка обнаружила три проблемы в данных:

- нарушения в стиле заголовков,
- пропущенные значения,
- дубликаты — явные и неявные.

Вы исправили заголовки, чтобы упростить работу с таблицей. Без дубликатов исследование станет более точным.

Пропущенные значения вы заменили на `'unknown'`. Ещё предстоит увидеть, не повредят ли исследованию пропуски в колонке `genre`.

Теперь можно перейти к проверке гипотез. 

## Проверка гипотез

### Сравнение поведения пользователей двух столиц

Первая гипотеза утверждает, что пользователи по-разному слушают музыку в Москве и Санкт-Петербурге. Проверьте это предположение по данным о трёх днях недели — понедельнике, среде и пятнице. Для этого:

* Разделите пользователей Москвы и Санкт-Петербурга
* Сравните, сколько треков послушала каждая группа пользователей в понедельник, среду и пятницу.


Для тренировки сначала выполните каждый из расчётов по отдельности. 

Оцените активность пользователей в каждом городе. Сгруппируйте данные по городу и посчитайте прослушивания в каждой группе.



In [18]:
# Подсчёт прослушиваний в каждом городе
df.groupby('city')['genre'].count()



city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64

<a id="error_2_2"></a> 
<div class="alert alert-danger">
    
<b>Ревьюер v2: ✋</b> Здесь нужно сгруппировать данные по городу воспользовавшись методом `groupby()`.
    
</div>

<a id="error_3_2"></a> 
<div class="alert alert-danger">
    
<b>Ревьюер v3: ✋</b> Между группировкой по столбцу `city` и вызовом метода `count()` нужно выбрать какой-нибудь из столбцов получившегося сгруппированного объекта, иначе в итоговой таблице будет несколько одинаковых столбцов.

Например: `df.groupby('day')['genre'].count()`
</div>

<div class="alert alert-success">

<b>Ревьюер v4: 👍</b> Отлично!

</div>

В Москве прослушиваний больше, чем в Петербурге. Из этого не следует, что московские пользователи чаще слушают музыку. Просто самих пользователей в Москве больше.

Теперь сгруппируйте данные по дню недели и подсчитайте прослушивания в понедельник, среду и пятницу. Учтите, что в данных есть информация только о прослушиваниях только за эти дни.


In [19]:
# Подсчёт прослушиваний в каждый из трёх дней
df.groupby('day')['genre'].count()
#Исправил

day
Friday       21840
Monday       21354
Wednesday    18059
Name: genre, dtype: int64

<a id="error_2_3"></a> 
<div class="alert alert-danger">
    
<b>Ревьюер v2: ✋</b> В этой ячейке тоже используй `groupby()`.
    
</div>

<a id="error_3_3"></a> 
<div class="alert alert-danger">
    
<b>Ревьюер v3: ✋</b> Между группировкой по столбцу `day` и вызовом метода `count()` нужно выбрать какой-нибудь из столбцов получившегося сгруппированного объекта, иначе в итоговой таблице будет несколько одинаковых столбцов.
    
</div>

<div class="alert alert-success">

<b>Ревьюер v4: 👍</b> Отлично v2!

</div>

В среднем пользователи из двух городов менее активны по средам. Но картина может измениться, если рассмотреть каждый город в отдельности.

Вы видели, как работает группировка по городу и по дням недели. Теперь напишите функцию, которая объединит два эти расчёта.

Создайте функцию `number_tracks()`, которая посчитает прослушивания для заданного дня и города. Ей понадобятся два параметра:
* день недели,
* название города.

В функции сохраните в переменную строки исходной таблицы, у которых значение:
  * в колонке `day` равно параметру `day`,
  * в колонке `city` равно параметру `city`.

Для этого примените последовательную фильтрацию с логической индексацией.

Затем посчитайте значения в столбце `user_id` получившейся таблицы. Результат сохраните в новую переменную. Верните эту переменную из функции.

In [20]:
# <создание функции number_tracks()>

# Объявляется функция с двумя параметрами: day, city.
def number_tracks(day, city):      
  
# В переменной track_list сохраняются те строки таблицы df, для которых 
# значение в столбце 'day' равно параметру day и одновременно значение
# в столбце 'city' равно параметру city (используйте последовательную фильтрацию
# с помощью логической индексации).
    track_list = df[(df['day'] == day) & (df['city'] == city)]
    
# В переменной track_list_count сохраняется число значений столбца 'user_id',
# рассчитанное методом count() для таблицы track_list.
    track_list_count = track_list['user_id'].count()
    
# Функция возвращает число - значение track_list_count.
    return track_list_count
 
#Функция для подсчёта прослушиваний для конкретного города и дня.
##сначала получит из исходной таблицы строки с нужным днём,
#затем из результата отфильтрует строки с нужным городом,
#методом count() посчитает количество значений в колонке user_id. 
#Это количество функция вернёт в качестве результата


<div class="alert alert-success">

<b>Ревьюер: 👍</b> Функция полностью выполняет поставленную перед ней задачу.  
Хорошо, что остался комментарий к своей функции. Это поможет не только другим аналитикам узнать, что она делает, но и тебе, спустя время вспомнить. Ниже я привел пример, как комментировать функции.  
```py
def number_tracks(day, city):

    """
    Функция для подсчёта прослушиваний для конкретного города и дня.
    С помощью последовательной фильтрации с логической индексацией она…
    """
    
    track_list = df[df['day'] == day]
    ...
```

</div>

Вызовите `number_tracks()` шесть раз, меняя значение параметров — так, чтобы получить данные для каждого города в каждый из трёх дней.

In [21]:
# количество прослушиваний в Москве по понедельникам
number_tracks('Monday', 'Moscow')

15740

In [22]:
# количество прослушиваний в Санкт-Петербурге по понедельникам
number_tracks('Monday', 'Saint-Petersburg')

5614

In [23]:
# количество прослушиваний в Москве по средам
number_tracks('Wednesday', 'Moscow')

11056

In [24]:
# количество прослушиваний в Санкт-Петербурге по средам
number_tracks('Wednesday', 'Saint-Petersburg')

7003

In [25]:
# количество прослушиваний в Москве по пятницам
number_tracks('Friday', 'Moscow')

15945

In [26]:
# количество прослушиваний в Санкт-Петербурге по пятницам
number_tracks('Friday', 'Saint-Petersburg')

5895

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Круто, данных для финального вывода становится больше. Пазл складывается в картинку!

</div>

Создайте c помощью конструктора `pd.DataFrame` таблицу, где
* названия колонок — `['city', 'monday', 'wednesday', 'friday']`;
* данные — результаты, которые вы получили с помощью `number_tracks`.

In [27]:
# Таблица с результатами
columns = ['city', 'monday', 'wednesday', 'friday']
data = [['Moscow', 15740, 11056, 15945],['Saint-Petersburg', 5614, 7003, 5895]]
table = pd.DataFrame(data = data, columns = columns)
display(table)
#таблица создана с учетом того, что дубликаты были удалены

city  monday  wednesday  friday
0            Moscow   15740      11056   15945
1  Saint-Petersburg    5614       7003    5895

<div class="alert alert-success">

<b>Ревьюер: 👍</b> Хорошо, что ты умеешь объявлять датафрейм, а еще что пишешь код, который удобно читать.

</div>

**Выводы**

Данные показывают разницу поведения пользователей:

- В Москве пик прослушиваний приходится на понедельник и пятницу, а в среду заметен спад.
- В Петербурге, наоборот, больше слушают музыку по средам. Активность в понедельник и пятницу здесь почти в равной мере уступает среде.

Значит, данные говорят в пользу первой гипотезы.

### Музыка в начале и в конце недели

Согласно второй гипотезе, утром в понедельник в Москве преобладают одни жанры, а в Петербурге — другие. Так же и вечером пятницы преобладают разные жанры — в зависимости от города.

Сохраните таблицы с данными в две переменные:
* по Москве — в `moscow_general`;
* по Санкт-Петербургу — в `spb_general`.

In [28]:
# получение таблицы moscow_general из тех строк таблицы df, 
# для которых значение в столбце 'city' равно 'Moscow'
moscow_general = df[df['city'] == 'Moscow']


In [29]:
# получение таблицы spb_general из тех строк таблицы df,
# для которых значение в столбце 'city' равно 'Saint-Petersburg'
spb_general = df[df['city'] == 'Saint-Petersburg']


<div class="alert alert-success">

<b>Ревьюер: 👍</b> Правильно. Логическая индексация в деле. 😉

</div>

Создайте функцию `genre_weekday()` с четырьмя параметрами:
* таблица (датафрейм) с данными,
* день недели,
* начальная временная метка в формате 'hh:mm', 
* последняя временная метка в формате 'hh:mm'.

Функция должна вернуть информацию о топ-10 жанров тех треков, которые прослушивали в указанный день, в промежутке между двумя отметками времени.

In [30]:
# Объявление функции genre_weekday() с параметрами table, day, time1, time2,
# которая возвращает информацию о самых популярных жанрах в указанный день в
# заданное время:
def genre_weekday(table, day, time1, time2):
    
# 1) в переменную genre_df сохраняются те строки переданного датафрейма table, для
#    которых одновременно: 
#    - значение в столбце day равно значению аргумента day
#    - значение в столбце time больше значения аргумента time1 
#    - значение в столбце time меньше значения аргумента time2
#    Используйте последовательную фильтрацию с помощью логической индексации.
    genre_df = table[(table['day']==day) & (table['time']>time1) & (table['time']<time2)]
    

# 2) сгруппировать датафрейм genre_df по столбцу genre, взять один из его
#    столбцов и посчитать методом count() количество записей для каждого из
#    присутствующих жанров, получившийся Series записать в переменную
#    genre_df_count
    genre_df_count = genre_df.groupby('genre')['genre'].count()
    
# 3) отсортировать genre_df_count по убыванию встречаемости и сохранить
#    в переменную genre_df_sorted
    genre_df_sorted = genre_df_count.sort_values(ascending = False)
# 4) вернуть Series из 10 первых значений genre_df_sorted, это будут топ-10
#    популярных жанров (в указанный день, в заданное время)
    return genre_df_sorted.head(10)


# Исправлено



<a id="error_2_4"></a>
<div class="alert alert-danger">

<b>Ревьюер: ✋</b> Между группировкой по столбцу и вызовом метода `count()` нужно выбрать какой-нибудь из столбцов получившегося сгруппированного объекта, иначе в итоговой таблице будет несколько одинаковых столбцов.
    
А поскольку группировку мы сделали по одному столбцу, то и аргумент `by = 'genre'` нужно убрать.
    
Подсказка:
```py
    genre_df_count = genre_df.groupby('genre')['genre'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending = False)
```

</div>

Cравните результаты функции `genre_weekday()` для Москвы и Санкт-Петербурга в понедельник утром (с 7:00 до 11:00) и в пятницу вечером (с 17:00 до 23:00):

In [31]:
# вызов функции для утра понедельника в Москве (вместо df — таблица moscow_general)
# объекты, хранящие время, являются строками и сравниваются как строки
# пример вызова: genre_weekday(moscow_general, 'Monday', '07:00', '11:00')
#Cравниваем полученные результаты по таблице для Москвы и Санкт-Петербурга в понедельник утром (с 7 до 11) и в пятницу вечером (с 17 до 23).
genre_weekday(moscow_general, 'Monday', '07:00:00', '11:00:00')

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: genre, dtype: int64

In [32]:
# вызов функции для утра понедельника в Петербурге (вместо df — таблица spb_general)
genre_weekday(spb_general, 'Monday', '07:00:00', '11:00:00')

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: genre, dtype: int64

In [33]:
# вызов функции для вечера пятницы в Москве
genre_weekday(moscow_general, 'Friday', '17:00:00', '23:00:00')

genre
pop            713
rock           517
dance          495
electronic     482
hiphop         273
world          208
ruspop         170
alternative    163
classical      163
rusrap         142
Name: genre, dtype: int64

In [34]:
# вызов функции для вечера пятницы в Петербурге
genre_weekday(spb_general, 'Friday', '17:00:00', '23:00:00')

genre
pop            256
rock           216
electronic     216
dance          210
hiphop          97
alternative     63
jazz            61
classical       60
rusrap          59
world           54
Name: genre, dtype: int64

**Выводы**

Если сравнить топ-10 жанров в понедельник утром, можно сделать такие выводы:

1. В Москве и Петербурге слушают похожую музыку. Единственное отличие — в московский рейтинг вошёл жанр “world”, а в петербургский — джаз и классика.

2. В Москве пропущенных значений оказалось так много, что значение `'unknown'` заняло десятое место среди самых популярных жанров. Значит, пропущенные значения занимают существенную долю в данных и угрожают достоверности исследования.

Вечер пятницы не меняет эту картину. Некоторые жанры поднимаются немного выше, другие спускаются, но в целом топ-10 остаётся тем же самым.

Таким образом, вторая гипотеза подтвердилась лишь частично:
* Пользователи слушают похожую музыку в начале недели и в конце.
* Разница между Москвой и Петербургом не слишком выражена. В Москве чаще слушают русскую популярную музыку, в Петербурге — джаз.

Однако пропуски в данных ставят под сомнение этот результат. В Москве их так много, что рейтинг топ-10 мог бы выглядеть иначе, если бы не утерянные  данные о жанрах.

### Жанровые предпочтения в Москве и Петербурге

Гипотеза: Петербург — столица рэпа, музыку этого жанра там слушают чаще, чем в Москве.  А Москва — город контрастов, в котором, тем не менее, преобладает поп-музыка.

Сгруппируйте таблицу `moscow_general` по жанру и посчитайте прослушивания треков каждого жанра методом `count()`. Затем отсортируйте результат в порядке убывания и сохраните его в таблице `moscow_genres`.

In [35]:
# одной строкой: группировка таблицы moscow_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в moscow_genres
moscow_genres = moscow_general.groupby('genre')['genre'].count().sort_values(ascending = False)
'''
почему в некоторых случаях достаточно написать genre_df.groupby('genre').count(), а в других как например 
в этой ячейке необходимо добавлять значение столбца в квадратных скобках?
'''

"\nпочему в некоторых случаях достаточно написать genre_df.groupby('genre').count(), а в других как например \nв этой ячейке необходимо добавлять значение столбца в квадратных скобках?\n"

<div class="alert alert-success">

<b>Ревьюер v4: 👍</b> Если бы мы искали сумму `sum()` или первое вхождение `first()` и так далее, то ненужно указывать столбец для группировки у нас бы вышла сгруппированная таблица с разными значениями. В принципе, подсчет количества `count()` тоже можно производить таким образом, но тогда у нас будет выходить таблица с одинаковыми значениями. В случае подсчета количества принято добавлять любой столбец, чтобы данные было удобнее воспринимать.


```py
df.groupby('city').first()
```

| city | user_id |	track |	artist | genre | time |	day |
|---|-----|-------|--------|-------|------|-----|
| Moscow |	55204538 |	Delayed Because of Accident |	Andreas Rönnberg |	rock |	14:07:09 |	Friday |
| Saint-Petersburg |	FFB692EC |	Kamigata To Boots |	The Mass Missile |	rock |	20:28:33 |	Wednesday |

```py
df.groupby('city').count()
```

| city | user_id |	track |	artist | genre | time |	day |
|---|-----|-------|--------|-------|------|-----|
| Moscow |	42741 | 42741 |	42741 |	42741 |	42741 |	42741 |
| Saint-Petersburg |	18512 | 	18512 |	18512 |	18512 | 18512 |	18512 |
    
</div>

Выведите на экран первые десять строк `moscow_genres`:

In [36]:
# просмотр первых 10 строк moscow_genres
moscow_genres.head(10)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> Данные для анализа предпочтений в Москве получены.

</div>

Теперь повторите то же и для Петербурга.

Сгруппируйте таблицу `spb_general` по жанру. Посчитайте прослушивания треков каждого жанра. Результат отсортируйте в порядке убывания и сохраните в таблице `spb_genres`:


In [37]:
# одной строкой: группировка таблицы spb_general по столбцу 'genre', 
# подсчёт числа значений 'genre' в этой группировке методом count(), 
# сортировка получившегося Series в порядке убывания и сохранение в spb_genres
spb_genres = spb_general.groupby('genre')['genre'].count().sort_values(ascending = False)

Выведите на экран первые десять строк `spb_genres`:

In [38]:
# просмотр первых 10 строк spb_genres
spb_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> По Питеру тоже данные есть.

</div>

**Выводы**

Гипотеза частично подтвердилась:
* Поп-музыка — самый популярный жанр в Москве, как и предполагала гипотеза. Более того, в топ-10 жанров встречается близкий жанр — русская популярная музыка.
* Вопреки ожиданиям, рэп одинаково популярен в Москве и Петербурге. 


## Итоги исследования

Вы проверили три гипотезы и установили:

1. День недели по-разному влияет на активность пользователей в Москве и Петербурге. 

Первая гипотеза полностью подтвердилась.

2. Музыкальные предпочтения не сильно меняются в течение недели — будь то Москва или Петербург. Небольшие различия заметны в начале недели, по понедельникам:
* в Москве слушают музыку жанра “world”,
* в Петербурге — джаз и классику.

Таким образом, вторая гипотеза подтвердилась лишь отчасти. Этот результат мог оказаться иным, если бы не пропуски в данных.

3. Во вкусах пользователей Москвы и Петербурга больше общего чем различий. Вопреки ожиданиям, предпочтения жанров в Петербурге напоминают московские.

Третья гипотеза не подтвердилась. Если различия в предпочтениях и существуют, на основной массе пользователей они незаметны.

**На практике исследования содержат проверки статистических гипотез.**
Из данных одного сервиса не всегда можно сделать вывод о всех жителях города.
Проверки статистических гипотез покажут, насколько они достоверны, исходя из имеющихся данных. 
С методами проверок гипотез вы ещё познакомитесь в следующих темах.

<div class="alert alert-success">

<b>Ревьюер v2: 👍</b> С итоговыми выводами все верно!

</div>

<div class="alert alert-danger">

<b>Ревьюер v2:</b> Проект еще не принят, осталось исправить несколько ошибок.
* Для удобства поиска я добавил тебе ссылки.

    * [Ответь на вопрос в пункте три](#error_2_1)
    * [Примени groupby() 1](#error_2_2)
    * [Примени groupby() 2](#error_2_3)
    * [Измени функцию](#error_2_4)
    
Жду правок 🙂

</div>

<div class="alert alert-danger">

<b>Ревьюер v3:</b> Андрей, осталось чуть-чуть и проект будет принят!

* [Ответь на вопрос](#error_3_1)
* [Вывод groupby() 1](#error_3_2)
* [Вывод groupby() 2](#error_3_3)
    
Жду правок 😉

</div>

<div class="alert alert-success">

<b>Ревьюер v4: ✌️</b> Андрей, упорства тебе не занимать, ты крут! А это залог успеха. 
    
Только увидел, что ты оставлял комментарии в коде. На вопросы ответил, и твой ответ на вопрос нашел. На будущее оставляй комментарии так чтобы их было видно.

Желаю тебе успехов в дальнейшем обучении и во всех твоих начинаниях, поздравляю **проект принят**! 🥳 🎉

</div>